Generate package-info files for all impl and mixin packages

Juuxel opened the pull request:
The generated files have `@ApiStatus.Internal` annotations for IDE warnings if mods try to access those classes directly.

Example output:

```java
// [License header here]

/**
 * Implementation code for fabric-loot-api-v2.
 */
@ApiStatus.Internal
package net.fabricmc.fabric.impl.loot;

import org.jetbrains.annotations.ApiStatus;
```

Juuxel committed:
Generate package-info files for all impl and mixin packages

Juuxel committed:
Add javadoc to generated package-infos, use multiline string

Juuxel committed:
Simplify code

Juuxel committed:
Remove manual ApiStatus.Internal on impl classes

Juuxel committed:
Update CONTRIBUTING.md to remove ApiStatus.Internal for impl classes

Juuxel committed:
Fix Gradle deprecations

Juuxel committed:
Fix more Gradle deprecations

Juuxel committed:
Add task metadata

modmuss50 reviewed (approved):
Great idea 👍 LGTM

modmuss50 commented on gradle/package-info.gradle:
```diff
@@ -0,0 +1,65 @@
+	outputDir = file('src/generated/java')
+}
+
+sourceSets.main.java.srcDir generateImplPackageInfos.outputDir
```
I think this might need splitting for the main/client source sets as its not going to be correctly denoting the package-infos for client only packages. Thus partly breaking splitting the jars in a mod dev env.

Juuxel replied:
Done without any workarounds now that #2617 has been merged

Juuxel committed:
Support client-main split

Juuxel committed:
Remove workaround for client packages

Juuxel committed:
Merge remote-tracking branch 'upstream/1.19.2' into package-info-gen

modmuss50 commented:
Please see build failuire.

Juuxel committed:
Try to fix Gradle 8 deprecations

Juuxel committed:
Try to fix Gradle 8 deprecations, part 2

I'm hoping that using the SourceDirectorySet instead of converting it to a
Set<File> (which is pointless?) will carry the build dependencies over.

Juuxel commented:
Fixed, the issue was the root build.gradle adding the module sources to `javadoc` but voiding the `Buildable` metadata by converting them to plain `Set<File>`s.

modmuss50 committed:
Add clean packageInfo task

modmuss50 reviewed (approved):
👍
